File: Farmsy/Features/Saved/SavedScreen.swift
import SwiftUI

/// Saved farms — the same favorites the user hearts on the website.
struct SavedScreen: View {
    var onOpenFarm: (FarmPin) -> Void

    @Environment(\.dismiss) private var dismiss
    @Environment(FarmsStore.self) private var farms
    @Environment(FavoritesStore.self) private var favorites
    @Environment(LocationManager.self) private var locationManager
    @Environment(SessionStore.self) private var session
    @Environment(\.requestAuth) private var requestAuth

    private var savedPins: [FarmPin] {
        let saved = farms.pins.filter { favorites.isSaved($0.osmId) }
        return farms.sortedByDistance(saved, from: locationManager.location)
    }

    var body: some View {
        VStack(spacing: 0) {
            // Same header treatment as the What's New sheet: an eyebrow and a
            // circular close.
            HStack {
                Text("SAVED FARMS")
                    .font(.geist(11, .semibold))
                    .kerning(1.2)
                    .foregroundStyle(Color.inkMuted)
                Spacer()
                Button { dismiss() } label: {
                    Image(systemName: "xmark")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Color(hex: 0x6B7280))
                        .frame(width: 32, height: 32)
                        .background(Color(hex: 0xF3F4F6), in: Circle())
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 16)
            .padding(.top, 16)
            .padding(.bottom, 8)

            content
        }
        .background(Color.cream.ignoresSafeArea())
    }

    @ViewBuilder
    private var content: some View {
        Group {
            if !session.isAuthenticated {
                VStack(spacing: 12) {
                    Spacer()
                    Text("🤍").font(.geist(54))
                    Text("Keep your favourites")
                        .font(.display(24))
                        .foregroundStyle(Color.ink)
                    Text("Sign in to save farms and find them here on every device.")
                        .font(.geist(15))
                        .foregroundStyle(Color.inkMuted)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 40)
                    Button("Sign in") {
                        Haptics.tap()
                        requestAuth()
                    }
                    .buttonStyle(PrimaryButtonStyle())
                    .padding(.horizontal, 60)
                    .padding(.top, 6)
                    Spacer()
                    Spacer()
                }
                .frame(maxWidth: .infinity)
            } else {
                // A numbered list, like the web: filled green circle + number for a
                // saved farm (with an X to remove), dashed circle + "Tap a heart to
                // save a farm" for the empty slots up to eight.
                let slots = max(8, savedPins.count)
                ScrollView(showsIndicators: false) {
                    VStack(spacing: 0) {
                        ForEach(0..<slots, id: \.self) { i in
                            if i < savedPins.count {
                                savedRow(index: i, pin: savedPins[i])
                            } else {
                                emptyRow(index: i)
                            }
                            if i < slots - 1 {
                                Divider().padding(.leading, 62)
                            }
                        }
                    }
                    .background(Color.creamCard, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous).stroke(Color.hairline, lineWidth: 1))
                    .padding(.horizontal, 14)
                    .padding(.top, 4)

                    // Discovery carousel — more farms worth saving.
                    TripRecommendations(onOpenFarm: { pin in dismiss(); onOpenFarm(pin) })
                        .padding(.horizontal, 14)
                        .padding(.top, 6)
                        .padding(.bottom, 16)
                }
            }
        }
    }

    private func savedRow(index: Int, pin: FarmPin) -> some View {
        HStack(spacing: 14) {
            Text("\(index + 1)")
                .font(.geist(13, .bold))
                .foregroundStyle(.white)
                .frame(width: 30, height: 30)
                .background(Color.farmGreenMap, in: Circle())
            VStack(alignment: .leading, spacing: 1) {
                Text(pin.name)
                    .font(.geist(15, .semibold))
                    .foregroundStyle(Color.ink)
                    .lineLimit(1)
                if let city = pin.city {
                    Text(city).font(.geist(12)).foregroundStyle(Color.inkMuted).lineLimit(1)
                }
            }
            Spacer(minLength: 6)
            Image(systemName: "xmark")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(Color.inkMuted)
                .frame(width: 32, height: 32)
                .contentShape(Rectangle())
                .tapCard { remove(pin) }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 11)
        .contentShape(Rectangle())
        .tapCard(excludeTopTrailing: 44) { onOpenFarm(pin) }
    }

    private func emptyRow(index: Int) -> some View {
        HStack(spacing: 14) {
            Text("\(index + 1)")
                .font(.geist(13, .bold))
                .foregroundStyle(Color.inkMuted.opacity(0.6))
                .frame(width: 30, height: 30)
                .overlay(Circle().strokeBorder(Color(hex: 0xE5E7EB), style: StrokeStyle(lineWidth: 1.5, dash: [3])))
            VStack(alignment: .leading, spacing: 2) {
                Text("Tap a heart to save a farm")
                    .font(.geist(15))
                    .foregroundStyle(Color.inkMuted)
                Text("—").font(.geist(12)).foregroundStyle(Color.inkMuted.opacity(0.5))
            }
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 11)
    }

    private func remove(_ pin: FarmPin) {
        guard let userId = session.session?.user.id else { return }
        Task { await favorites.toggle(pin.osmId, userId: userId) }
    }
}

File: android/app/src/main/java/app/farmsy/android/features/saved/SavedScreen.kt
package app.farmsy.android.features.saved

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.LocalFarms
import app.farmsy.android.LocalFavorites
import app.farmsy.android.LocalLocationHelper
import app.farmsy.android.LocalRequestAuth
import app.farmsy.android.LocalSession
import app.farmsy.android.R
import app.farmsy.android.core.FarmPin
import app.farmsy.android.features.discover.RecommendationCarousel
import app.farmsy.android.ui.theme.FarmsyColors
import app.farmsy.android.ui.theme.Kicker
import app.farmsy.android.ui.theme.PrimaryButton
import app.farmsy.android.ui.theme.display
import app.farmsy.android.ui.theme.geist
import kotlinx.coroutines.launch

/// Saved farms — 1:1 with iOS SavedScreen: an eyebrow header, then a numbered
/// slot list (filled green circle + name + city + X-remove for a saved farm,
/// dashed circle + "Tap a heart to save a farm" for the empty slots up to eight),
/// with the recommendation carousel below. Guest state invites sign-in.
@Composable
fun SavedScreen(onOpenFarm: (FarmPin) -> Unit) {
    val session = LocalSession.current
    val farms = LocalFarms.current
    val favorites = LocalFavorites.current
    val location = LocalLocationHelper.current
    val requestAuth = LocalRequestAuth.current
    val scope = rememberCoroutineScope()

    val currentSession by session.session.collectAsState()
    val pins by farms.pins.collectAsState()
    val savedIds by favorites.osmIds.collectAsState()
    val loc by location.location.collectAsState()

    val saved = farms.sortedByDistance(pins.filter { savedIds.contains(it.osmId) }, loc)

    Column(Modifier.fillMaxSize().background(FarmsyColors.cream)) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Kicker(stringResource(R.string.saved_farms))
        }

        if (currentSession == null) {
            Column(
                Modifier.fillMaxSize().padding(40.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
            ) {
                Text("🤍", fontSize = 54.sp)
                Spacer(Modifier.height(12.dp))
                Text(stringResource(R.string.keep_your_favourites), style = display(24.sp), color = FarmsyColors.ink)
                Spacer(Modifier.height(6.dp))
                Text(
                    stringResource(R.string.sign_in_to_save_farms_and_find_them_here_on_every_device),
                    style = geist(15.sp), color = FarmsyColors.inkMuted, textAlign = TextAlign.Center,
                )
                Spacer(Modifier.height(16.dp))
                PrimaryButton(stringResource(R.string.sign_in), Modifier.padding(horizontal = 20.dp), onClick = requestAuth)
            }
        } else {
            val slots = maxOf(8, saved.size)
            Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState())) {
                Column(
                    Modifier.padding(horizontal = 14.dp, vertical = 4.dp)
                        .background(FarmsyColors.creamCard, RoundedCornerShape(16.dp))
                        .border(1.dp, FarmsyColors.hairline, RoundedCornerShape(16.dp)),
                ) {
                    for (i in 0 until slots) {
                        if (i < saved.size) {
                            SavedRow(i, saved[i], onOpen = { onOpenFarm(saved[i]) }, onRemove = {
                                val uid = currentSession?.user?.id ?: return@SavedRow
                                scope.launch { favorites.toggle(saved[i].osmId, uid) }
                            })
                        } else {
                            EmptyRow(i)
                        }
                        if (i < slots - 1) HorizontalDivider(Modifier.padding(start = 62.dp), color = FarmsyColors.hairline)
                    }
                }
                RecommendationCarousel(
                    onOpenFarm = onOpenFarm,
                    modifier = Modifier.padding(horizontal = 14.dp).padding(top = 6.dp, bottom = 16.dp),
                )
            }
        }
    }
}

@Composable
private fun SavedRow(index: Int, pin: FarmPin, onOpen: () -> Unit, onRemove: () -> Unit) {
    Row(
        Modifier.fillMaxWidth().clickable { onOpen() }.padding(horizontal = 12.dp, vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Box(Modifier.size(30.dp).background(FarmsyColors.farmGreenMap, CircleShape), contentAlignment = Alignment.Center) {
            Text("${index + 1}", style = geist(13.sp, FontWeight.Bold), color = Color.White)
        }
        Column(Modifier.weight(1f)) {
            Text(pin.name, style = geist(15.sp, FontWeight.SemiBold), color = FarmsyColors.ink, maxLines = 1)
            pin.city?.let { Text(it, style = geist(12.sp), color = FarmsyColors.inkMuted, maxLines = 1) }
        }
        Icon(
            Icons.Filled.Close, null, tint = FarmsyColors.inkMuted,
            modifier = Modifier.size(32.dp).clickable { onRemove() }.padding(9.dp),
        )
    }
}

@Composable
private fun EmptyRow(index: Int) {
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Box(
            Modifier.size(30.dp).border(1.5.dp, FarmsyColors.inkMuted.copy(alpha = 0.3f), CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Text("${index + 1}", style = geist(13.sp, FontWeight.Bold), color = FarmsyColors.inkMuted.copy(alpha = 0.6f))
        }
        Column(Modifier.weight(1f)) {
            Text(stringResource(R.string.tap_a_heart_to_save_a_farm), style = geist(15.sp), color = FarmsyColors.inkMuted)
            Text("—", style = geist(12.sp), color = FarmsyColors.inkMuted.copy(alpha = 0.5f))
        }
    }
}
